freqtrade/strategy/strategy_helper.py

```diff
@@ -83,3 +85,57 @@ def stoploss_from_open(open_relative_stop: float, current_profit: float) -> floa
 
     # negative stoploss values indicate the requested stop price is higher than the current price
     return max(stoploss, 0.0)
+
+
+def get_trade_candle(dataframe, trade: Trade, timeframe: str, now: datetime = None):
+    """
+    search for nearest row of trade.open_date
+    """
+    trade_candle = find_candle_datetime(dataframe,
+                                        timeframe,
+                                        query_date=trade.open_date_utc,
+                                        pair=trade.pair,
+                                        now=now)
+    return trade_candle
+
+
+def get_buy_candle(dataframe, trade: Trade, timeframe: str, now: datetime = None):
+    """
+    search for nearest row of trade.open_date minus 1 candle (the buy decision candle)
+    """
+    trade_open = trade.open_date_utc
+    one_frame = timedelta(minutes=timeframe_to_minutes(timeframe))
+    trade_candle = find_candle_datetime(dataframe,
+                                        timeframe,
+                                        query_date=trade_open - one_frame,
+                                        pair=trade.pair,
+                                        now=now)
+    return trade_candle
+
+
+def find_candle_datetime(dataframe: pd.DataFrame, timeframe: str, query_date: datetime, pair: str, now: datetime = None):
+    result = None
+    candle = find_candle_datetime_safer(dataframe, query_date)
+    # candle = find_candle_datetime_faster(dataframe, timeframe, query_date, now)
+    result = candle if candle.empty else candle.squeeze()
+    return result
+
+
+def find_candle_datetime_faster(dataframe: pd.DataFrame, timeframe: str, query_date: datetime, now: datetime = None):
+    if(now and now == query_date):
+        candle = dataframe.iloc[-1]
+    else:
+        candle_date = timeframe_to_prev_date(self.timeframe, query_date)
+        candle = dataframe.loc[dataframe.date == candle_date]
+    return candle
+
+
+def find_candle_datetime_safer(dataframe: pd.DataFrame, query_date: datetime):
+    df = dataframe[['date']].set_index('date')
+
+    try:
+        date_mask = df.index.unique().get_loc(query_date, method='ffill')
+        candle = dataframe.iloc[date_mask]  # use iloc because date_mask maybe :int
+    except KeyError:  # trade.open_date may not exist yet
+        candle = pd.DataFrame(index=dataframe.index)
+    return candle
```

Comment:
Considering the enhancements @rokups made for on "lookahead bias prevention" - i'm not sure i do see a problem with the "faster" approach - although the speed difference you measured is not big enough to really matter (considering this will only be called for "open trade candles" in backtesting) which way we use it.


Also, if you look at [dataframe access](https://www.freqtrade.io/en/latest/strategy-advanced/#dataframe-access) in the docs - it's using .iloc as well (thanks to the enhancements mentioned above).

The only real danger here (but with both of these methods) is if they are used out of a `populate*()` method - as neither of the 2 versions can really be used for that (you'll not have a current time - and these functions should be populated / used in a vectorized way).

As such - i think we should decide on one - but i'm not sure on the benefits of the "safer" variant ... although i think the exception handler might be wrong `dataframe.index` will return a series with 500+ rows ... but the expectation from this is 1 row i'd think.

Reply:
Also `find_candle_datetime_faster` would produce invalid results when we pass `current_time` as `now` parameter without passing it through `timeframe_to_prev_date()`. But even if that was fixed this function is incorrect, because it would return different candles depending on current time.

I do not understand `find_candle_datetime_safer` at all. Seems like it is doing ``candle = dataframe.loc[dataframe.date == candle_date]` except with a dataframe copy, and on error it returns empty dataframe, while on success it returns a `Series` object.